R/Design-methods.R

```diff
@@ -4710,3 +4710,311 @@ setMethod("simulate",
+              samples = this_samples,
+              model = object@model,
+              data = this_combo_data,
+              group = factor("combo", levels = c("mono", "combo"))
+            )
+
+            stop_combo_results <- h_unpack_stopit(stop_combo)
+
+            if (object@same_dose) {
+              this_mono_dose <- this_combo_dose <- min(
+                this_mono_dose,
+                this_combo_dose
+              )
+            }
+          }
+        }
+
+        fit_mono <- fit(
+          object = this_samples,
+          model = object@model,
+          data = grouped_data,
+          group = factor("mono", levels = c("mono", "combo"))
+        )
+        fit_combo <- fit(
+          object = this_samples,
+          model = object@model,
+          data = grouped_data,
+          group = factor("combo", levels = c("mono", "combo"))
+        )
+
+        this_result <- list(
+          mono = list(
+            data = this_mono_data,
+            dose = this_mono_dose,
+            fit = subset(fit_mono, select = -dose),
+            stop = attr(stop_mono, "message"),
+            report_results = stop_mono_results
+          ),
+          combo = list(
+            data = this_combo_data,
+            dose = this_combo_dose,
+            fit = subset(fit_combo, select = -dose),
+            stop = attr(stop_combo, "message"),
+            report_results = stop_combo_results
+          )
+        )
+
+        return(this_result)
+      }
+
+      result_list <- getResultList(
+        fun = run_sim,
+        nsim = nsim,
+        vars =
+          c(
+            "simSeeds",
+            "args",
+            "nArgs",
+            "truth",
+            "combo_truth",
+            "firstSeparate",
+            "object",
+            "mcmcOptions"
+          ),
+        parallel = if (parallel) nCores else NULL
+      )
+
+      # Now we have a list with each element containing mono and combo,
+      # but we want it now the other way around, i.e. a list with 2 elements
+      # mono and combo and the iterations inside.
+      result_list <- list(
+        mono = lapply(result_list, "[[", "mono"),
+        combo = lapply(result_list, "[[", "combo")
+      )
+
+      # Put everything in a list with both mono and combo Simulations:
+      lapply(result_list, function(this_list) {
+        data_list <- lapply(this_list, "[[", "data")
+        recommended_doses <- as.numeric(sapply(this_list, "[[", "dose"))
+        fit_list <- lapply(this_list, "[[", "fit")
+        stop_reasons <- lapply(this_list, "[[", "stop")
+        report_results <- lapply(this_list, "[[", "report_results")
+        stop_report <- as.matrix(do.call(rbind, report_results))
+
+        Simulations(
+          data = data_list,
+          doses = recommended_doses,
+          fit = fit_list,
+          stop_reasons = stop_reasons,
+          stop_report = stop_report,
+          seed = rng_state
+        )
+      })
+    }
+)
```

Comment:
At over 250 lines, this function is loooong!  Can we simplify with helper functions and lists?  One obvious possibility is to use a helper to gather results from either mono or combo simulations.  For example,

```
this_dose <- lapply(
  c("mono, "combo"),
  function(x) {
    nextBest(
              slot(object, x)@nextBest,
              doselimit = dose_limit[[x]],
              samples = this_samples,
              model = object@model,
              data = grouped_data,
              group = factor(x, levels = c("mono", "combo"))
            )$value
  }
)
```

and similar elsewhere.

That could potentially halve the length of the function.

Reply:
haha :D yes indeed! let me see what I can do now. we might also get more shortening later when doing general refactoring of simulate methods in #671